pkg/backup/backupdest/backup_index.go

```diff
@@ -138,6 +142,169 @@ func IndexExists(ctx context.Context, store cloud.ExternalStorage, subdir string
 	return indexExists, nil
 }
 
+// ListIndexes lists all the index files for a backup chain rooted by the full
+// backup indicated by the subdir. The store should be rooted at the default
+// collection URI (the one that contains the `index/` directory). It returns
+// the basenames of the listed index files. It assumes that the subdir is
+// resolved and not `LATEST`.
+//
+// The indexes are returned in reverse chronological order, i.e. the most recent
+// index is first in the list.
+func ListIndexes(
+	ctx context.Context, store cloud.ExternalStorage, subdir string,
+) ([]string, error) {
+	var indexBasenames []string
+	if err := store.List(
+		ctx,
+		indexSubdir(subdir)+"/",
+		"",
+		func(file string) error {
+			indexBasenames = append(indexBasenames, path.Base(file))
+			return nil
+		},
+	); err != nil {
+		return nil, errors.Wrapf(err, "listing indexes in %s", subdir)
+	}
+	return indexBasenames, nil
+}
+
+// GetBackupTreeIndexMetadata concurrently retrieves the index metadata for all
+// backups within the specified subdir, up to the specified end time, inclusive.
+// It also returns the total size of memory allocated to store the index files.
+// The returned indexes are sorted in reverse chronological order, i.e. the
+// latest index is first in the list. The store should be rooted at the
+// collection URI that contains the `index/` directory.
+//
+// Note: Even if an error is returned, the total memory reserved in the monitor
+// is still returned as it may be non-zero.
+func GetBackupTreeIndexMetadata(
+	ctx context.Context,
+	mem *mon.BoundAccount,
+	store cloud.ExternalStorage,
+	subdir string,
+	endTime hlc.Timestamp,
+) ([]backuppb.BackupIndexMetadata, int64, error) {
+	indexBasenames, err := ListIndexes(ctx, store, subdir)
+	if err != nil {
+		return nil, 0, err
+	}
+	// Note that the following logic depends on the fact that ListIndexes
+	// returns the index files in reverse chronological order.
+	startIdx := 0
+	if !endTime.IsEmpty() {
+		// We will still need to open each of the index files to ensure that the
+		// backup end times (which are nano-second specific) are within the
+		// specified endTime, but we can do some preliminary filtering first based
+		// on the filename.
+		for startIdx = len(indexBasenames) - 1; startIdx >= 0; startIdx-- {
+			_, indexEnd, err := parseIndexFilename(indexBasenames[startIdx])
+			if err != nil {
+				return nil, 0, err
+			}
+			// GoTime always truncates an hlc.Timestamp, so to ensure we don't
+			// accidentally leave out an index, we wait until the truncated index end
+			// time exceeds the truncated hlc.Timestamp end time.
+			if indexEnd.After(endTime.GoTime()) {
+				break
+			}
+		}
+		indexBasenames = indexBasenames[max(startIdx-1, 0):]
+	}
+
+	indexes := make([]backuppb.BackupIndexMetadata, len(indexBasenames))
+	memMu := struct {
```

Comment:
after thinking about this more, i think we can scrap the memory monitor. Imagine each file is 1KB (which is a large overestimate). If we read 400 of these files, 400kb is not worth memory monitoring.